tests/e2e/v3_curl_maxstream_test.go

```diff
@@ -109,31 +105,23 @@ func testV3CurlMaxStream(t *testing.T, reachLimit bool, opts ...ctlOption) {
 		expectedResponse = "Operation timed out"
 	}
 	wg.Add(int(concurrentNumber))
-	t.Logf("Running the test, MaxConcurrentStreams: %d, concurrentNumber: %d, expectedResponse: %s\n",
+	t.Logf("Running the test, MaxConcurrentStreams: %d, concurrentNumber: %d, expected range's response: %s\n",
 		cx.cfg.MaxConcurrentStreams, concurrentNumber, expectedResponse)
-	errCh := make(chan error, concurrentNumber)
-	submitConcurrentWatch(cx, int(concurrentNumber), &wg, errCh)
-	submitRangeAfterConcurrentWatch(cx, expectedResponse)
 
-	// Step 4: check the watch errors. Note that we only check the watch error
-	// before closing cluster. Once we close the cluster, the watch must run
-	// into error, and we should ignore them by then.
-	t.Log("Checking watch error.")
-	select {
-	case werr := <-errCh:
-		t.Fatal(werr)
-	default:
-	}
+	closeServerCh := make(chan struct{})
+	submitConcurrentWatch(cx, int(concurrentNumber), &wg, closeServerCh)
+	submitRangeAfterConcurrentWatch(cx, expectedResponse)
 
-	// Step 5: Close the cluster
+	// Step 4: Close the cluster
 	t.Log("Closing test cluster...")
+	close(closeServerCh)
```

Comment:
It isn't correct to close the channel directly, because `submitConcurrentWatch`  might not finish the task yet.

Reply:
If I understand correctly, the `submitConcurrentWatch` is used to increase the active streams before hiting the quota, which depends on the case. If the case wants `Range` request fails with timeout, the `submitConcurrentWatch` will hit the quota of `MaxConcurrentStream`. Otherwise, the `submitConcurrentWatch` will left one availiable stream to `Range` request. 

So, in this commit, the `submitConcurrentWatch` can make sure that all the active streams are ready before return. And the `submitRangeAfterConcurrentWatch` will verify that the `MaxConcurrentStreams` quota setting is working. After `submitRangeAfterConcurrentWatch` finish, it is ready to close the server and let all the watch connections closes.( It is not attempt to verify the watch event, and the `Range` request doesn't create event)

But the watch connection might exit somehow, like oom killer. It is unexpected. So I use a channel `closeServerCh` to mark that the server is ready to close. If the watch connection is close and the `closeServerCh` is still open, it should fail.

```go
                       // hold the connection and wait for server shutdown
			perr := proc.Close()

			// curl process will return
			select {
			case <-closeCh:
			default:
				// perr could be nil.
				return fmt.Errorf("unexpected connection close before server closes: %v", perr)
			}
			return nil
```

Does it make sense to you?

Reply:
I missed the info that `submitRangeAfterConcurrentWatch` has already been called before closing the channel.  It makes sense.